### DISCUSSION TITLE
Realtime not working if JWT is set using supabase.auth.setAuth

### DISCUSSION THREAD

**Discussion by gopal1996** [supabase/supabase#8737] [Questions] [answered]
Hi, I am doing POC to test the supabase feature, If I tried to Authenticate Supabase with JWT, the real-time connection is not working. but if I Commented on that part, then real-time is working.

I am generating JWT token with following payload
```python
token = jwt.encode({
            "Account_id": "Ac4Nd5Ovrkuj",
            "User_id": "Us4TFmyZaTz5",
            "role": "anon"
        },
        app.config["SECRET_KEY"])
```

Generated Token passed to Supabase js client code

```javascript
supabase.auth.setAuth(access_token);

const data = supabase
  .from("kissflow")
  .on("*", (payload) => {
    console.log("Change received!", payload);
  })
  .subscribe();
```

After running the above snippet, getting the below error in the web socket
<img width="934" alt="image" src="https://user-images.githubusercontent.com/22369081/187615685-3311a15c-b9c0-4fd3-adee-fa99e33aaee4.png">

Realtime connection is in joining state
<img width="997" alt="image" src="https://user-images.githubusercontent.com/22369081/187615860-0c14ac76-123d-46a2-a40e-bec4059e53e4.png">

nothing is returned if I run the following command
<img width="821" alt="image" src="https://user-images.githubusercontent.com/22369081/187616107-fa2dee5e-f2b1-492d-a1bd-2726757d316f.png">

But If I disabled `setAuth` code, everything is working as normal
<img width="976" alt="image" src="https://user-images.githubusercontent.com/22369081/187616282-0e1b7f44-f1ef-488e-a52c-0f8a8b2a9b74.png">

<img width="847" alt="image" src="https://user-images.githubusercontent.com/22369081/187617016-17d8c4d6-4de7-40f4-879d-ca3072b200f5.png">

Even If I use the same access token to retrieve the data from the database it works perfectly
```javascript
async function getData() {
  let userId = "Us4TFmyZaTz5";
  const { user, error } = supabase.auth.setAuth(access_token);
  const { data } = await supabase.from("kissflow").select(`
  payload -> ${userId}
  `);

  return data;
}
getData().then((res) => console.log(res));
```
The above snippet is working, the only issue is with the real-time database

can you please help us with what is going wrong with this? 

Tried below steps also mentioned in the below issues
[#6290](https://github.com/supabase/supabase/issues/6290)
[#7440](https://github.com/supabase/supabase/issues/7440)

**Comment by GaryAustin1** [comment, marked as answer]
I'm going to guess you are missing a claim that a standard Supabase jwt has and realtime requires.  In a quick search I did not find a list of what a Supabase instance has that set to.  You might experiment and add a few of the other claims to see what works.... if this is the issue....

Note there is the below also, but I don't think Supabase hosting sets it.
Realtime server has a config option for the jwt info it requires.
`JWT_CLAIM_VALIDATORS       # {string}      Expected claim key/value pairs compared to JWT claims via equality checks in order to validate JWT. e.g. '{"iss": "Issuer", "nbf": 1610078130}'. This is optional but encouraged.`

**Reply by GaryAustin1** [reply]
In particular you do not show an exp claim.  I think that would be needed.

**Reply by gopal1996** [reply]
once "exp" is added, it is working

**Comment by KTibow** [comment]
Having the same problem. I just tested in a JSFiddle and when I use `setAuth` with my own token, it doesn't work.
Update: Fixed by setting `"role": "authenticated"`
